File: packages/core/src/tools/pg-schema.ts
import { z } from "zod";
import { ActionRegistry, ActionContext, ToolDefinition } from "../types.js";
import { listHandler } from "../actions/schema/list.js";
import { describeHandler } from "../actions/schema/describe.js";
import { ddlHandler } from "../actions/schema/ddl.js";

const schemaRegistry: ActionRegistry = {
    list: listHandler,
    describe: describeHandler,
    create: ddlHandler,
    alter: ddlHandler,
    drop: ddlHandler,
};

export const PgSchemaToolSchema = z.object({
    action: z.enum(["list", "describe", "create", "alter", "drop"]).describe("Action to perform on schema objects"),
    target: z.enum(["database", "schema", "table", "column", "index", "view", "function", "trigger", "sequence", "constraint"]).describe("Target object type"),
    schema: z.string().optional().describe("Schema name (filter for list, or target schema for others)"),
    name: z.string().optional().describe("Object name (required for describe, create, alter, drop)"),
    definition: z.string().optional().describe("SQL definition body (for create/alter, e.g. column defs or view query)"),
    table: z.string().optional().describe("Parent table name (filter for list triggers/constraints)"),
    session_id: z.string().optional().describe("Session ID for transactional operations"),
    autocommit: z.boolean().optional().describe("Execute DDL immediately (required if no session_id provided)"),
    options: z.object({
        // List options
        include_sizes: z.boolean().optional().describe("Include table sizes in list output"),
        include_materialized: z.boolean().optional().describe("Include materialized views when listing views"),
        limit: z.number().optional().describe("Max results to return (list)"),
        offset: z.number().optional().describe("Pagination offset (list)"),
        // DDL options
        cascade: z.boolean().optional().describe("Use CASCADE to drop dependent objects (drop)"),
        if_exists: z.boolean().optional().describe("Add IF EXISTS (drop)"),
        if_not_exists: z.boolean().optional().describe("Add IF NOT EXISTS (create)"),
    }).optional().describe("Options for list or DDL operations"),
});

export async function pgSchemaHandler(params: any, context: ActionContext) {
    const handler = schemaRegistry[params.action];
    if (!handler) {
        throw new Error(`Unknown action: ${params.action}`);
    }
    return await handler.handler(params, context);
}

export const pgSchemaTool: ToolDefinition = {
    name: "pg_schema",
    config: {
        description: `Manage database structure and schema objects (DDL).

Actions:
  • list: Enumerate schemas, tables, views, functions, triggers, sequences, constraints (read-only)
  • describe: Get detailed structure of tables (columns, indexes) (read-only)
  • create: Create new tables, indexes, views (requires session_id OR autocommit:true)
  • alter: Modify existing tables (requires session_id OR autocommit:true)
  • drop: Remove objects with optional CASCADE (requires session_id OR autocommit:true)

Safety: DDL mutations (create/alter/drop) use Default-Deny policy.
Without session_id or autocommit:true, DDL will fail with a safety error.

Examples:
  {"action": "list", "target": "table"}
  {"action": "describe", "target": "table", "name": "users"}
  {"action": "create", "target": "table", "name": "logs", "definition": "id SERIAL PRIMARY KEY", "autocommit": true}`,
        inputSchema: PgSchemaToolSchema,
        destructiveHint: true, // Contains create/alter/drop actions that can modify schema
    },
    handler: (context) => (params) => pgSchemaHandler(params, context),
};

File: coldquery/tools/pg_schema.py
from typing import Literal
from fastmcp.exceptions import ToolError
from coldquery.dependencies import CurrentActionContext
from coldquery.core.context import ActionContext
from coldquery.app import mcp
from coldquery.actions.schema.list import list_handler
from coldquery.actions.schema.describe import describe_handler
from coldquery.actions.schema.ddl import create_handler, alter_handler, drop_handler

SCHEMA_ACTIONS = {
    "list": list_handler,
    "describe": describe_handler,
    "create": create_handler,
    "alter": alter_handler,
    "drop": drop_handler,
}

@mcp.tool()
async def pg_schema(
    action: Literal["list", "describe", "create", "alter", "drop"],
    target: str | None = None,  # table, view, schema, function, trigger, sequence, constraint
    name: str | None = None,
    schema: str | None = None,
    sql: str | None = None,
    limit: int = 100,
    offset: int = 0,
    include_sizes: bool = False,
    cascade: bool = False,
    if_exists: bool = False,
    if_not_exists: bool = False,
    session_id: str | None = None,
    autocommit: bool | None = None,
    context: ActionContext = CurrentActionContext(),
) -> str:
    """Manage database schema and introspection.

    Actions:
    - list: List database objects (tables, views, functions, etc.)
    - describe: Get detailed structure of a table
    - create: Create database objects with DDL
    - alter: Modify existing database objects
    - drop: Remove database objects
    """
    handler = SCHEMA_ACTIONS.get(action)
    if not handler:
        raise ToolError(f"Unknown action: {action}")

    params = {
        "target": target,
        "name": name,
        "schema": schema,
        "sql": sql,
        "limit": limit,
        "offset": offset,
        "include_sizes": include_sizes,
        "cascade": cascade,
        "if_exists": if_exists,
        "if_not_exists": if_not_exists,
        "session_id": session_id,
        "autocommit": autocommit,
    }

    return await handler(params, context)
